### DISCUSSION TITLE
Add the ability to configure the camera entity WebRTC Streams

### DISCUSSION THREAD

**Discussion by sdb9696** [home-assistant/architecture#1040] [Entity Models]
**Context**

The camera integration currently provides [WebRTC Streams](https://developers.home-assistant.io/docs/core/entity/camera/#webrtc-streams) to implementers by allowing them to override `async_handle_web_rtc_offer` which takes an SDP offer from HA and returns an answer.  The HA front end then makes a peer connection to the stream provider.

Generally the steps in setting up a web rtc stream on the client side are:

1. Create an RTC peer connection by providing an [RtcConfiguration](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/RTCPeerConnection) for the connection (This can be empty).
2. Set up transceivers for audio and video.
3. Generate an [SDP offer](https://en.wikipedia.org/wiki/Session_Description_Protocol) and send it to the broker.

The broker/integration then sends back the SDP answer and the stream should start.

Currently the camera implementation does not allow a RtcConfiguration to be provided for the step 1. or control over how the transceivers are setup for step 2. The audio and video transceivers are set up as recvonly on the client and if the connection requires additional information such as stun servers to be provided in the configuration this is not possible.

Finally the broker between RTC peers may need to clean up resources after the session ends.  For example if it has to keep pinging the provider source to keep it from disconnecting during the session.  This is not currently possible.

N.B. It appears the configuration requirement might have been an issue for some implementers of RTPS to WebRTC, as a [PR]( https://github.com/home-assistant/core/pull/72574) went in while ago to allow setting a single stun server for implementers of rtsp_to_webrtc.

**Proposal**

The proposal is to allow WebRTC implementers to override two additional methods `async_get_web_rtc_config` and `async_handle_web_rtc_close` for the HA front end to call, such that:

1. RtcConfiguration can be provided for the connection, including the ability to provide multiple stun/turn servers.
2. Transceiver direction can be set allowing two way voice communication.  i.e. when someone rings on a doorbell it can be answered in HA directly without having to open a vendor application.
3. The broker/integration will have the option to clean up any resources it may be using.

It is not expected that any configuration should be exposed to HA end users, rather it would be provided by the implementing integration.

Users should have the ability to mute their microphone.

The implementation should be backwards compatible so that integrations currently overriding `async_handle_web_rtc_offer` get the same behaviour as currently.  (I believe that at the moment in core this is only the nest integration).

I have a draft [core](https://github.com/home-assistant/core/pull/110489) and [frontend](https://github.com/home-assistant/frontend/pull/19792) PR for anyone who may want to see an implementation.  This has been tested and works with ring cameras on chrome (chrome requires ssl for microphone access so audio is one way otherwise)

**Consequences**

- More integrations should be able to implement WebRTC.
- Integrations should be able to utilise two way audio in WebRTC streams if needed.

**Related discussions**

I've reviewed the related discussions from when WebRTC was implemented https://github.com/home-assistant/architecture/discussions/640.  The rtsp_to_webrtc discussion is also related https://github.com/home-assistant/architecture/discussions/658.

**Comment by allenporter** [comment]
Appreciate the detailed architecture proposal. This generally seems like a reasonable proposal (edit: reasonable feature to support, rather). Camera has a decent amount of complexity already however, so I am curious to understand the use planned use cases.

What core integrations are going to adopt this? There is an example for rtsp_to_webrtc stun server settings, though given that already has a one-off solution for part of the proposal and the PR doesn't seem to moving that to the new approach, so i assume that isn't the target for this use case. Nest does not support two way audio in the API as far as i can tell. So those are the only two core integrations I am aware of using this.

**Reply by allenporter** [reply]
A few additional comments/corrections on my previous commment:
- I read further in the PR and see that this adds webrtc support for ring, so nice to see additional uses of webrtc by other integrations.  Great.
- rtsp_to_webrtc should be updated to use this mechanism for passing stun servers.
- I would prefer to be more explicit about which configuration options are included/allowed (my preference would be things from the spec are added as needed/used only)
- It would be good to talk through the rationale for "configuration" as an entity. Do configuration options need to change per entity? or can they be per configuration entry?


One problem with the "stop" call or doing any cleanup is that Home Assistant may have many clients of the frontend viewing a stream at once. In order to do propper cleanup, it implies that either the frontend needs some notion of a session token to identify which stream to close. Otherwise, if you stop one frontend stream it may break another frontend stream.

FWIW Nest has an opposite challenge now of needing to keep alive a stream that expires (right now it just fails after the stream expires, though i was considering some kind of retry to just reopen). Perhaps some shared notion of a client keepalive though could help with either keeping open or closing a stream.

**Reply by sdb9696** [reply]
Thanks for the input, some replies below.

> * rtsp_to_webrtc should be updated to use this mechanism for passing stun servers.

Makes sense although I'd need your advice about how best to do it.  Would you see it going in a subsequent PR or with the initial changes?

> * I would prefer to be more explicit about which configuration options are included/allowed (my preference would be things from the spec are added as needed/used only)

I agree and I initially kept it to the RtcConfig options to those I thought ring needed, but as it turns out it seems the bundle_policy and Ice transport policy defaults work for chrome so I'm not sending them.  That said it could be a client thing and IOS/android defaults might be different.  For the transport direction the video_direction could be dropped for now.

> * It would be good to talk through the rationale for "configuration" as an entity. Do configuration options need to change per entity? or can they be per configuration entry?

For the ring use case I'm able to test, the configuration is currently a config entry level thing.  However that's because I don't have devices that don't support video or two way audio etc so I think it should be a entity level configuration to be future proof and cover more use cases.

> One problem with the "stop" call or doing any cleanup is that Home Assistant may have many clients of the frontend viewing a stream at once. In order to do propper cleanup, it implies that either the frontend needs some notion of a session token to identify which stream to close. Otherwise, if you stop one frontend stream it may break another frontend stream.

So the ring_doorbell library is acting as a kind of session manager for the session.  It initiates the session with the ring api over a websocket and gets a session id.  Once it's brokered the session with ICE and SDP it has to send a ping every 3 seconds with the session_id and gets a pong back, otherwise it gets closed by the ring api with a reason of "ping not recevied". So the close here just stops pinging and closes the socket so I don't think it would affect other sessions.

> FWIW Nest has an opposite challenge now of needing to keep alive a stream that expires (right now it just fails after the stream expires, though i was considering some kind of retry to just reopen). Perhaps some shared notion of a client keepalive though could help with either keeping open or closing a stream.

Does nest integration have a similar concept of a session manager in the integration's library?  If so it'd be interesting to see if the protocol messages are similar because AFAICT the ring one is proprietary but perhaps not.

>  Nest does not support two way audio in the API as far as i can tell. So those are the only two core integrations I am aware of using this.

Does it support it in the native app?  For ring I didn't need to do anything api wise, just on the peer connection and it worked (edit: with chrome secure connection check supressed).

**Comment by allenporter** [comment]
> Thanks for the input, some replies below.
> 
> > * rtsp_to_webrtc should be updated to use this mechanism for passing stun servers.
> 
> Makes sense although I'd need your advice about how best to do it. Would you see it going in a subsequent PR or with the initial changes?

Wasn't really thinking about PRs yet..  just that we need to ensure the proposal can and will replace existing one offs.  (subsequent PRs is fine)

> > * It would be good to talk through the rationale for "configuration" as an entity. Do configuration options need to change per entity? or can they be per configuration entry?
> 
> For the ring use case I'm able to test, the configuration is currently a config entry level thing. However that's because I don't have devices that don't support video or two way audio etc so I think it should be a entity level configuration to be future proof and cover more use cases.

Yes, OK, I agree that different devices have different capabilities. 

We should think through a few alternative approaches.

Here are some thoughts:
- The frontend needs to know the capabilities of the devices to:
    - Change how the UI looks, for example to render a microphone button
    - Change options on the webrtc client (for example, setup additional tracks)
- That is, the configuration is really a straight pass through here. It is tempting to pass webrtc configuration data to support new use cases, but I think its a pretty broad API and most changes seem like they would still probably require frontend changes anyway
- I don't think WebRTC necessarily needs its own API for communicating device capabilities. I think the minimum needed to support ring is effectively a boolean to flip the audio receive enum value to support directional audio
- We don't need to use a single mechanism for ice server configuration and for device capabilities. The stun server config feels more like configuration entity specific configuration data to me.

I'd like to see what a "minimum" constrained API for this could look like, and I think we should separate the stun server config from the per device capabilities could also help.

> 
> > One problem with the "stop" call or doing any cleanup is that Home Assistant may have many clients of the frontend viewing a stream at once. In order to do propper cleanup, it implies that either the frontend needs some notion of a session token to identify which stream to close. Otherwise, if you stop one frontend stream it may break another frontend stream.
> 
> So the ring_doorbell library is acting as a kind of session manager for the session. It initiates the session with the ring api over a websocket and gets a session id. Once it's brokered the session with ICE and SDP it has to send a ping every 3 seconds with the session_id and gets a pong back, otherwise it gets closed by the ring api with a reason of "ping not recevied". So the close here just stops pinging and closes the socket so I don't think it would affect other sessions.

Let me spell out an example to see if we're talking about the same thing. (I'm not sure if we are).  I am thinking about a case where you have two users in the same home both viewing Home Assistant.
- Alice opens Home Assistant mobile app, navigates to a frontend card displaying "Camera #1" live stream (webrtc session is opened and is streaming)
- Bob opens Home Assistant mobile app, navigates to a frontend card displaying "Camera #1"  live stream (webrtc session is opened and is streaming)
- Blob switches to another tab, closing the frontend card displaying the camera. Bob's webrtc session with "Camera #1" is closed.

Expected result: Alice's web stream with "Camera #1" should still keep playing.

It sounds like this proposal assumes some home assistant core session management needs to happen though we haven't established how the frontend and core communicate about this.

> 
> > FWIW Nest has an opposite challenge now of needing to keep alive a stream that expires (right now it just fails after the stream expires, though i was considering some kind of retry to just reopen). Perhaps some shared notion of a client keepalive though could help with either keeping open or closing a stream.
> 
> Does nest integration have a similar concept of a session manager in the integration's library? If so it'd be interesting to see if the protocol messages are similar because AFAICT the ring one is proprietary but perhaps not.

There is a gap in handling extending streams today, and I was hesitant to add more complexity to home assistant core to handle this given Nest is the only real large user of the APIs. From what you describes above, I think we can work together on a single solution to support both Ring and Nest here.  See the existing issue for [Nest supporting extending streams](https://github.com/home-assistant/core/issues/86610).   From what you describes, the requirements are basically the same but Ring has a 3 second timeout where Nest has a 5 minute timeout. [Nest WebRTC API](https://developers.google.com/nest/device-access/api/camera-battery#access_a_live_stream) just has another call to extend the stream, keeping it alive.

> 
> > Nest does not support two way audio in the API as far as i can tell. So those are the only two core integrations I am aware of using this.
> 
> Does it support it in the native app? For ring I didn't need to do anything api wise, just on the peer connection and it worked (edit: with chrome secure connection check supressed).

Yes, native app can do this however the SDM api (link above) explicitly states `The audio setting must be set to a=recvonly (receive only).`

So in summary my recommendation for this proposal would be we explore:
- Alternatives where we can use existing APIs for communicating device capabilities rather than webrtc config on the entity.
- Alternatives for the ice server configuration (e.g. per configuration entity)
- Explore session keep-alive / close proposal that can support both Ring and Nest

I'm happy to also chat in on home assistant discord if that makes it faster to work through some design proposal ideas.

**Reply by sdb9696** [reply]
Thanks Allen, I have reached out on Discord so we can continue the discussion.

**Comment by OnFreund** [comment]
While we're extending the interface between frontend and backend for WebRTC, it would be great to also add the ability to send candidates asynchronously. The way it's set up today, the backend has to wait for all the candidates before it can send them over.
Might not be in scope for this proposal, but something to consider.
